dev/io.openliberty.org.jboss.resteasy.common/src/org/jboss/resteasy/core/ServerResponseWriter.java

```diff
@@ -0,0 +1,752 @@
+         boolean hasStarStar = false;
+         boolean hasApplicationStar = false;
+         boolean pFound = false;
+         for (MediaType accept : accepts) {
+            //Instead of getting all the MBWs compatible with type/generic and then filtering using accept, we use
+            //getPossibleMessageBodyWritersMap to get the viable MBWs for the given type AND accept.
+            Map<MessageBodyWriter<?>, Class<?>> mbws = providerFactory.getPossibleMessageBodyWritersMap(type, generic, annotations, accept);
+            for (Entry<MessageBodyWriter<?>, Class<?>> e : mbws.entrySet())
+            {
+               MessageBodyWriter<?> mbw = e.getKey();
+               Class<?> wt = e.getValue();
+               Produces produces = mbw.getClass().getAnnotation(Produces.class);
+               if (produces == null) produces = WILDCARD_PRODUCES; // Spec, section 4.2.3 "Declaring Media Type Capabilities"
+               for (String produceValue : produces.value())
+               {
+                  pFound = true;
+                  MediaType produce = MediaType.valueOf(produceValue);
+
+                  if (produce.isCompatible(accept))
+                  {
+                     SortableMediaType ms = mostSpecific(produce, wt, accept, null);
+                     if (ms.isWildcardSubtype())
+                     {
+                        hasStarStar |= ms.isWildcardType();
+                        hasApplicationStar |= ms.getType().equals("application");
+                     }
+                     M.add(ms);
+                  }
+               }
+            }
+         }
+         if (!pFound) // JAX-RS 2.0 Section 3.8.3
+         {
+            for (MediaType accept : accepts)
+            {
+               MediaType produce = MediaType.WILDCARD_TYPE;
+               if (produce.isCompatible(accept))
+               {
+                  SortableMediaType ms = mostSpecific(produce, null, accept, null);
+                  if (ms.isWildcardSubtype())
+                  {
+                     hasStarStar |= ms.isWildcardType();
+                     hasApplicationStar |= ms.getType().equals("application");
+                  }
+                  M.add(ms);
+               }
+            }
+         }
+         chosen = chooseFromM(chosen, M, hasStarStar, hasApplicationStar);
+      }
+      
+   // Liberty change begins
+      if(chosen.getParameters().containsKey(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES) ||
+                      chosen.getParameters().containsKey("q") || chosen.getParameters().containsKey("qs")) {
+          Map<String, String> map = new HashMap<String, String>(chosen.getParameters());
+          map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES);
+          map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES_LC);
+          map.remove("q");
+          map.remove("qs");
+          chosen = new MediaType(chosen.getType(), chosen.getSubtype(), map);
+      }
```

Comment:
```suggestion
      Map<String, String> map = null;
      if (chosen.getParameters().containsKey(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES)) {
         map = new HashMap<String, String>(chosen.getParameters());
         map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES);
         map.remove(SegmentNode.RESTEASY_SERVER_HAS_PRODUCES_LC);
      }

      if (chosen.getParameters().containsKey("q")) {
    	  if (map == null) {
 	         map = new HashMap<String, String>(chosen.getParameters());
    	  }
          map.remove("q");
      }

      if (chosen.getParameters().containsKey("qs")) {
    	  if (map == null) {
 	         map = new HashMap<String, String>(chosen.getParameters());
    	  }
          map.remove("qs");
      }

      if (map != null) {
         chosen = new MediaType(chosen.getType(), chosen.getSubtype(), map);
      }
```

Reply:
The original code was similar to your suggestion in checking for the key before calling remove (though your code improves on the original by creating only one HashMap and one MediaType). But the cost of remove should be very similar to the cost of containsKey, for a key that is not there, because remove would first try to find the item, and if it is not found, then its work is done. I ran this theory by Vijay and he concurred in principle, though he had not looked into the JDK code yet. 

So I came up with an implementation that pulled chosen.getParameters into the new HashMap first, unconditionally, did all 4 removes on it (with no containsKey calls), and if any of the removes came back non-null, created the new MediaType with the modified HashMap. I didn't like creating the HashMap all the time, but in the hasProduces case, we were gonna create it anyway ... maybe I should have stuck with that version. 

Anyway, I came up with this compromise - by stringing the containsKey calls together with || then execution will skip any after the first hit, so the number of containsKey calls is potentially reduced and there is at most one HashMap and MediaType created. 

I don't feel strongly about this ... none of the approaches I've seen so far ticks all the boxes.